sql/sql_class.h

```diff
@@ -1566,6 +1567,66 @@ typedef struct
   ulonglong start_utime;
 } QUERY_START_TIME_INFO;
 
+/**
+   A single-hash-function bloom filter for approximate accessed page
+   counter.
+*/
+class Bloom_filter {
+ private:
+  /** Bloom filter size */
+  enum { SIZE= 8192 };
+
+  typedef std::bitset<SIZE> Bit_set;
+
+  /** The bit set, which is allocated in a MEM_ROOT */
+  Bit_set *bit_set;
+
+ public:
+  Bloom_filter() : bit_set(NULL) {}
+
+  ~Bloom_filter()
+  {
+    // Do not delete, just destruct, due to MEM_ROOT allocation
+    bit_set->~bitset();
+  }
+
+  void clear()
+  {
+    // Do not delete, just force new MEM_ROOT allocation on the next use
+    bit_set= NULL;
+  }
+
+  /**
+     Check whether key is maybe a member of a set
+
+     @param[in, out]    mem_root        MEM_ROOT to allocate the bit set in, if
+      not allocated already
+     @param[in]         key             key whose presence to check
+
+     @return if true, the key might be a member of the set. If false, the key
+     is definitely not a member of the set.
+  */
+  bool test_and_set(MEM_ROOT *mem_root, ulong key)
```

Comment:
Shouldn't `mem_root` be a constructor's parameter?

Reply:
Saves a pointer